stacks-signer/src/v0/signer_state.rs

```diff
@@ -544,4 +760,433 @@ impl LocalStateMachine {
 
         Ok(())
     }
+
+    /// Updates the local state machine's viewpoint as necessary based on the global state
+    pub fn capitulate_viewpoint(
+        &mut self,
+        signerdb: &mut SignerDb,
+        eval: &mut GlobalStateEvaluator,
+        local_address: StacksAddress,
+    ) {
+        // Before we ever access eval...we should make sure to include our own local state machine update message in the evaluation
+        let local_update: Result<StateMachineUpdateMessage, _> = (&*self).try_into();
+        let Ok(mut local_update) = local_update else {
+            return;
+        };
+
+        let old_protocol_version = local_update.active_signer_protocol_version;
+        // First check if we should update our active protocol version
+        let active_signer_protocol_version = eval
+            .determine_latest_supported_signer_protocol_version(local_address, &local_update)
+            .unwrap_or(old_protocol_version);
+
+        let StateMachineUpdateContent::V0 {
+            burn_block,
+            burn_block_height,
+            current_miner,
+            ..
+        } = &local_update.content;
+
+        if active_signer_protocol_version != old_protocol_version {
+            info!("Updating active signer protocol version from {old_protocol_version} to {active_signer_protocol_version}");
+            *self = Self::Initialized(SignerStateMachine {
+                burn_block: *burn_block,
+                burn_block_height: *burn_block_height,
+                current_miner: current_miner.into(),
+                active_signer_protocol_version,
+            });
+            // Because we updated our active signer protocol version, update local_update so its included in the subsequent evaluations
+            let update: Result<StateMachineUpdateMessage, _> = (&*self).try_into();
+            let Ok(update) = update else {
+                return;
+            };
+            local_update = update;
+        }
+
+        // Check if we should also capitulate our miner viewpoint
+        let Some(new_miner) = eval.capitulate_miner_view(signerdb, local_address, &local_update)
+        else {
+            return;
+        };
+
+        let StateMachineUpdateContent::V0 {
+            burn_block,
+            burn_block_height,
+            current_miner,
+            ..
+        } = local_update.content;
+
+        if current_miner != new_miner {
+            info!("Capitulating local state machine's current miner viewpoint";
+                "current_miner" => ?current_miner,
+                "new_miner" => ?new_miner,
+            );
+            *self = Self::Initialized(SignerStateMachine {
+                burn_block,
+                burn_block_height,
+                current_miner: (&new_miner).into(),
+                active_signer_protocol_version,
+            });
+        }
+    }
+}
+
+/// Tests for SignerDb
+#[cfg(test)]
+mod tests {
```

Comment:
Can you move these to a different file? Once we get to the ~1k line range for a file, I think we should be separating the test module out.